### PR TITLE
Feature/bbs 254 sso

### PR DESCRIPTION
**Pull request by xendk**
Implements login via island.is.

### CODE REVIEW

File: deploy.php [hunk 1]
```diff
@@ -280,6 +282,7 @@
   'drush:updb',
   'drush:css_generate',
   'drush:ccall',
+  'drush:updb',
```
Comment (kasperg) [inline deploy.php]: Why run `updb` twice?

File: modules/aleph/src/Aleph/AlephClient.php [hunk 2]
```diff
@@ -86,57 +86,64 @@ public function __construct($base_url, $base_url_rest, $catalog_library, $item_l
...
+        return new \SimpleXMLElement($response->getBody());
+      }
 
-    // Throw exception if the status from Aleph is not OK.
-    throw new AlephClientException($response->error, $response->code);
+      // Throw exception if the status from Aleph is not OK.
+      throw new AlephClientException($response->error, $response->code);
+    }
+    catch (AlephClientException $e) {
+      watchdog_exception('aleph', $e, t('Aleph returned a non 200 HTTP status coe'), [], WATCHDOG_ALERT);
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: The control flow seems odd there.

As far as I can tell L109 is the only place where an `﻿﻿﻿﻿﻿﻿AlepthClientException` is thrown. It is then immediatly caught on L111. `$this->client->request()` on L101 will throw GuzzleExceptions which are not caught.

Also: How do we signal to callers that an error has occurred if we catch exceptions internally? The DocBlock has not been updated to reflect this.

Nitpicking but there is a typo here.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 3]
```diff
@@ -86,57 +86,64 @@ public function __construct($base_url, $base_url_rest, $catalog_library, $item_l
...
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: See comments above.

File: modules/aleph/src/Aleph/AlephClient.php [hunk 4]
```diff
@@ -504,4 +493,41 @@ public function borByKey(AlephPatron $patron) {
     return $response;
   }
 
+
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: Nitpicking but additional whitespace here?

File: modules/aleph/src/Aleph/AlephClient.php [hunk 5]
```diff
@@ -504,4 +493,41 @@ public function borByKey(AlephPatron $patron) {
...
+    $options = [];
+
+    $xml = new \SimpleXMLElement('<get-new-patron></get-new-patron>');
+    $new_patron = $xml->addChild('new-patron');
+    $new_patron
+      ->addChild('patron-name', $patron->getName())
+      ->addChild('internal-id', $patron->getId())
+      // Local sub library is hardcoded to BBAAA,
+      // as all new users should be there.
+      ->addChild('local-sub-library', 'BBAAA')
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: Is there any way we can avoid hardcoding this?

File: modules/aleph/src/Aleph/AlephClient.php [hunk 6]
```diff
@@ -504,4 +493,41 @@ public function borByKey(AlephPatron $patron) {
...
+      ->addChild('internal-id-verification', $patron->getVerification())
+      ->addChild('email-address', $patron->getEmail())
+      // The other value is GER (german), so we hard-code english.
+      ->addChild('con-lng', 'ENG')
```
Comment (kasperg) [inline modules/aleph/src/Aleph/AlephClient.php]: :+1:

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php [hunk 7]
```diff
@@ -42,6 +42,22 @@ public function __construct(AlephClient $client, AlephPatron $patron = NULL) {
     $this->patron = $patron;
   }
 
+  /**
+   * Get patron by $ssn.
+   */
+  public function getPatronByName($ssn) {
+    $patron = new AlephPatron();
+    $patron->setId($ssn);
```
Comment (kasperg) [inline modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: It seems odd that a method called `getPatronByName` takes `$ssn` as an argument which again is used as the patron id.

How about `getPatronBySsn()` or `getPatronById()?

If we use abbreviations like ssn then please write them out in the DocBlock.

File: modules/ding_innskraning/ding_innskraning.install [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+ * @file
+ * Install, update and uninstall functions for the ding_innskraning module.
+ */
+
+/**
+ * Implements hook_install().
+ */
+function ding_innskraning_install() {
+  // Needed by ding_innskraning_form_profile2_form_alter() to ensure it's called
+  // after ding_user_form_profile2_form_alter().
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.install:13]: :+1:

File: modules/ding_innskraning/ding_innskraning.module [hunk 9]
```diff
@@ -0,0 +1,780 @@
...
+    'title' => 'Ding Innskraning settings',
+    'description' => 'Settings for Ding Innskraning module',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('ding_innskraning_settings'),
+    'access arguments' => array('administer ding innskraning'),
+    'type' => MENU_NORMAL_ITEM,
+    'file' => 'inc/ding_innskraning.admin.inc',
+  );
+
+  $items['saml/login'] = array(
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module:45]: Nitpicking but I would normally expected that modules "namespaced" their menu entries e.g. `ding_innskraning/saml/login`.

File: modules/ding_innskraning/ding_innskraning.module [hunk 10]
```diff
@@ -0,0 +1,780 @@
...
+        'lib/Saml2/Auth.php',
+        'lib/Saml2/AuthnRequest.php',
+        'lib/Saml2/Constants.php',
+        'lib/Saml2/Error.php',
+        'lib/Saml2/LogoutRequest.php',
+        'lib/Saml2/LogoutResponse.php',
+        'lib/Saml2/Metadata.php',
+        'lib/Saml2/Response.php',
+        'lib/Saml2/Settings.php',
+        'lib/Saml2/Utils.php',
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: Can we autoload these using `xautoload`?

File: modules/ding_innskraning/ding_innskraning.module [hunk 11]
```diff
@@ -0,0 +1,780 @@
...
+ *   The state of the form.
+ */
+function ding_innskraning_form_alter(array &$form, array &$form_state, $form_id) {
+  $login_forms = array(
+    'user_login',
+    'user_login_block',
+    'ding_user_authenticate_form',
+  );
+  if (in_array($form_id, $login_forms)) {
+    dpm($form);
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: No `dpm` in VCS please.

File: modules/ding_innskraning/ding_innskraning.module [hunk 12]
```diff
@@ -0,0 +1,780 @@
...
+function ding_innskraning_saml_login() {
+  libraries_load('xmlseclibs');
+
+  $provider_url = 'https://innskraning.island.is';
+  $url = $provider_url;
+  $id = variable_get('ding_innskraning_id', '');
+
+  // Start a session to store the token in, so we can check it on return. This
+  // guards against attackers redirecting victims to island.is with return urls
+  // that might cause trouble. Currently the damage might be limited (delting
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: :+1:

Nitpicking: `deleting`.

File: modules/ding_innskraning/ding_innskraning.module [hunk 13]
```diff
@@ -0,0 +1,780 @@
...
+  $parameters['authid'] = $token;
+  // Create the request body.
+  $samlRequest = _ding_innskraning_login_request($token, $id, $provider_url);
+  $parameters['SAMLRequest'] = $samlRequest;
+  $parameters['RelayState'] = url('<front>', array('absolute' => TRUE));
+  // @TODO: why is this commented out.
+  // Sign the request.
+  /*$signature = _ding_innskraning_signature($samlRequest, $parameters['RelayState']);
+    $parameters['SigAlg'] = XMLSecurityKey::RSA_SHA1;
+    $parameters['Signature'] = $signature;*/
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module:197]: I agree that keeping this in the codebase is OK.

File: modules/ding_innskraning/ding_innskraning.module [hunk 14]
```diff
@@ -0,0 +1,780 @@
...
+
+/**
+ * Consume authentications from https://innskraning.island.is.
+ */
+function ding_innskraning_saml_endpoint() {
+  // Extract token from request.
+  if (!array_key_exists('token', $_POST)) {
+    watchdog('ding_innskraning', "Post parameter 'token' not found in SAML response.", array(), WATCHDOG_WARNING);
+    drupal_goto('user/login');
+    return FALSE;
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: return `FALSE` after `drupal_goto()`?

File: modules/ding_innskraning/ding_innskraning.module [hunk 15]
```diff
@@ -0,0 +1,780 @@
...
+
+  // Load XML document.
+  $dom = new DOMDocument();
+  $dom->loadXML($xml);
+
+  if (!$dom) {
+    drupal_set_message(t('Authentication failed'), 'error');
+    watchdog('ding_innskraning', 'SAML Response could not be processed, XML: @xml', array('@xml' => $xml), WATCHDOG_ERROR);
+    drupal_goto('user/login');
+    return FALSE;
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: Again.

File: modules/ding_innskraning/ding_innskraning.module [hunk 16]
```diff
@@ -0,0 +1,780 @@
...
+ * Creates an id for the authentication request. island.is requires the id to be
+ * formatted as a GUID.
+ */
+function _ding_innskraning_auth_id() {
+  // Implementation created from
+  // https://www.texelate.co.uk/blog/create-a-guid-with-php
+  // Create a token.
+  $token = $_SERVER['HTTP_HOST'];
+  $token .= $_SERVER['REQUEST_URI'];
+  $token .= uniqid(rand(), TRUE);
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: Is it intentional that this does not use a more secure source of random data like suggested in the linked blogpost?

File: modules/ding_innskraning/ding_innskraning.module [hunk 17]
```diff
@@ -0,0 +1,780 @@
...
+
+  // GUID is 128-bit hex.
+  $hash = strtoupper(md5($token));
+
+  // Create formatted GUID.
+  $guid = '';
+
+  // GUID format is XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX for readability.
+  $guid .= substr($hash, 0, 8) . '-' . substr($hash, 8, 4) . '-' . substr($hash, 12, 4) .
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: Why append to an empty string instead of just setting it?

File: modules/ding_innskraning/ding_innskraning.module [hunk 18]
```diff
@@ -0,0 +1,780 @@
...
+ * Create the login request with SAML parameters.
+ */
+function _ding_innskraning_login_request($id, $issuer, $destination) {
+  // Add the destination URL for login.
+  $destination = url($destination, array('query' => array('id' => $issuer, 'auth' => $id)));
+  $destination = htmlspecialchars($destination);
+  // Take the name of the site as provider name.
+  $provider = variable_get('site_name', '');
+  // Log the time when request was created.
+  $issueInstant = format_date(time(), 'custom', "Y-m-dTH:i:sZ");
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: Should this not be `REQUEST_TIME`?

File: modules/ding_innskraning/ding_innskraning.module [hunk 19]
```diff
@@ -0,0 +1,780 @@
...
+    <samlp:RequestedAuthnContext Comparison="exact">
+        <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:X509</saml:AuthnContextClassRef>
+    </samlp:RequestedAuthnContext>
+</samlp:AuthnRequest>
+AUTHNREQUEST;
+
+  $deflatedRequest = gzdeflate($request);
+  $base64Request = base64_encode($deflatedRequest);
+
+  return $base64Request;
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module:330]: This incantation already seems to be handled by [`AuthnRequest` from `php-saml`](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/AuthnRequest.php). Why is this not used?

File: modules/ding_innskraning/ding_innskraning.module [hunk 20]
```diff
@@ -0,0 +1,780 @@
...
+
+  if ($name) {
+    return $file_location;
+  }
+
+  if ($file_location != '') {
+    $key = file_get_contents($file_location);
+  }
+
+  return $key;
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: I do not understand this method. It seems to return the path to the key in one branch and the contents of the key in others.

File: modules/ding_innskraning/ding_innskraning.module [hunk 21]
```diff
@@ -0,0 +1,780 @@
...
+}
+
+/**
+ * Sign the SAML request.
+ */
+function _ding_innskraning_signature($samlRequest, $relayState, $signAlgorithm = XMLSecurityKey::RSA_SHA1) {
+  $key = _ding_innskraning_get_key();
+  if (!$key) {
+    drupal_set_message(t('No key provided, authentication can not be finished.'), 'info');
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: This is not a error message suitable for end users. They have no clue about what a key is. If anything this should be handled in a `hook_requirements()` implementations.

File: modules/ding_innskraning/ding_innskraning.module [hunk 22]
```diff
@@ -0,0 +1,780 @@
...
+
+  // Create new account in Drupal and if one exists update it.
+  $account = _ding_user_create_account($ssn, $auth_res);
+
+  // Check that a profile exists for the user and if not create one.
+  _ding_user_create_profile2($account);
+
+  $form_state['uid'] = $account->uid;
+  user_login_submit(array(), $form_state);
+  return TRUE;
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: Why return anything here? The function cannot return anything but `TRUE`.

File: modules/ding_innskraning/ding_innskraning.module [hunk 23]
```diff
@@ -0,0 +1,780 @@
...
+  // Validate XML file structure.
+  $schemaFile = libraries_get_path('php-saml') . '/lib/Saml2/schemas/saml-schema-protocol-2.0.xsd';
+  $oldEntityLoader = libxml_disable_entity_loader(FALSE);
+  $res = $xmlDoc->schemaValidate($schemaFile);
+  libxml_disable_entity_loader($oldEntityLoader);
+  if (!$res) {
+    $xmlErrors = libxml_get_errors();
+    watchdog('ding_innskraning', 'Error validating the metadata: @vars', array('@vars' => var_export($xmlErrors, TRUE)));
+    throw new Exception('Error validating the metadata');
+  }
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module:485]: It seems as if much of this validation could be handled by [`Response` for `php-saml`](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Response.php#L100).

File: modules/ding_innskraning/ding_innskraning.module [hunk 24]
```diff
@@ -0,0 +1,780 @@
...
+  if (!$objKeyInfo->key && empty($key)) {
+    $key = _ding_innskraning_get_key(TRUE);
+    $objKey->loadKey($key, TRUE);
+  }
+
+  if (!$objXMLSecDSig->verify($objKey)) {
+    throw new Exception("Signature invalid!");
+  }
+  else {
+    return "Signature valid.";
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: This returns a `string`? That seems very odd.

File: modules/ding_innskraning/ding_innskraning.module [hunk 25]
```diff
@@ -0,0 +1,780 @@
...
+  }
+
+  $serialNumber = $parsed['subject']['serialNumber'];
+  $issuer = $parsed['issuer']['CN'];
+
+  if ($serialNumber != "6503760649") {
+    throw new Exception("Not the correct serial number in the certificate");
+  }
+
+  if ($issuer != "Traustur bunadur") {
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module:708]: There are some hardcoded values here. Could they be extracted?

File: modules/ding_innskraning/ding_innskraning.module [hunk 26]
```diff
@@ -0,0 +1,780 @@
...
+    throw new Exception("Not correct CA");
+  }
+
+  return TRUE;
+}
+
+/**
+ * Get client IP.
+ */
+function _ding_innskraning_get_client_ip() {
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module]: Why use this instead of Drupals own [`ip_address()`](https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/ip_address/7.x)?

File: modules/ding_innskraning/test/SAML_response.xml [hunk 27]
```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kasperg) [inline modules/ding_innskraning/test/SAML_response.xml]: Why is this included? Consinder explaining this somewhere - perhaps in a comment at the beginning of this file?

File: modules/ding_innskraning/ding_innskraning.module [hunk 28]
```diff
@@ -0,0 +1,780 @@
...
+ */
+function _ding_innskraning_verify($xmlDoc) {
+  global $base_url;
+
+  // Load library.
+  libraries_load('xmlseclibs');
+
+  // Check SAML version.
+  if ($xmlDoc->documentElement->getAttribute('Version') != '2.0') {
+    throw new Exception('Unsupported SAML version');
```
Comment (kasperg) [inline modules/ding_innskraning/ding_innskraning.module:444]: I am not a fan of throwing the generic exception but that seems to be the general pattern here 😕 .

**Review by kasperg — CHANGES_REQUESTED** [review]
Questions and comments ahoy! 🚢

File: deploy.php (hunk 1 above)
Comment (xendk) [reply deploy.php]: Gah, merge booboo.

File: modules/aleph/src/Aleph/AlephClient.php (hunk 2 above)
Comment (xendk) [reply modules/aleph/src/Aleph/AlephClient.php]: I donno.

I have no clear idea of where error handling is supposed to be. @tripox?

But I'd guess that `request` and `requestRest` should be `protected` and the other methods in ` AlephClient` should handle their exceptions?

File: modules/aleph/src/Aleph/AlephClient.php (hunk 4 above)
Comment (xendk) [reply modules/aleph/src/Aleph/AlephClient.php]: fixed

File: modules/aleph/src/Aleph/AlephClient.php (hunk 5 above)
Comment (xendk) [reply modules/aleph/src/Aleph/AlephClient.php]: @tripox ?

File: modules/aleph/src/Aleph/Handler/AlephPatronHandler.php (hunk 7 above)
Comment (xendk) [reply modules/aleph/src/Aleph/Handler/AlephPatronHandler.php]: Spillover from Ding which refers to name rather than CPR.

I'll change it to `getPatronById()`.

File: modules/ding_innskraning/ding_innskraning.module (hunk 9 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module:45]: Yeah, but menu entries is also publicly visible, so often they're not. But `saml` aren't good either.

I've changed it to `innskraning/login`.

Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module:45]: Scratch that. island.is is hardcoded to `saml/consume`, so for consistency..

File: modules/ding_innskraning/ding_innskraning.module (hunk 10 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: done

File: modules/ding_innskraning/ding_innskraning.module (hunk 11 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: I know, I know...

File: modules/aleph/src/Aleph/AlephClient.php (hunk 5 above)
Comment (tripox) [reply modules/aleph/src/Aleph/AlephClient.php]: This could be a setting in `Aleph provider` instead.

File: modules/ding_innskraning/ding_innskraning.module (hunk 14 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Hah, I was blinded by the returns after throws. All removed.

File: modules/aleph/src/Aleph/AlephClient.php (hunk 2 above)
Comment (tripox) [reply modules/aleph/src/Aleph/AlephClient.php]: You got it, @xendk

File: modules/ding_innskraning/ding_innskraning.module (hunk 16 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Well, https://tools.ietf.org/html/rfc4122#section-6:

> Do not assume that UUIDs are hard to guess; they should not be used as security capabilities (identifiers whose mere possession grants access), for example. 

Think it's a bit overkill, but I've changed it to use `drupal_random_bytes` which uses `openssl_random_pseudo_bytes` if available.

File: modules/ding_innskraning/ding_innskraning.module (hunk 17 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Fixed

File: modules/ding_innskraning/ding_innskraning.module (hunk 19 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module:330]: I do not know. That was how the module was implemented, and I was tasked with getting it to work, not re-implement it.

There might be a reason for not using AuthnRequest or there might not be, hard to say without a history of the module.

File: modules/ding_innskraning/ding_innskraning.module (hunk 20 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Yeah that's stupid. I'll split it.

Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Gah, and it seems to be confused with cert.

File: modules/ding_innskraning/ding_innskraning.module (hunk 21 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Yes, but it's not used (the code calling it is commented).

File: modules/ding_innskraning/ding_innskraning.module (hunk 22 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Fixed.

File: modules/ding_innskraning/ding_innskraning.module (hunk 23 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module:485]: Yeah, it would seem so...

File: modules/ding_innskraning/ding_innskraning.module (hunk 24 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Fixed

File: modules/ding_innskraning/ding_innskraning.module (hunk 25 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module:708]: To where? It's the way it's done in the official documentation: https://www.island.is/media/pdf-skjol-a-island.is-2014/Island.is-IAS-services-Instructions-vs2.0_4_july_2014.pdf

File: modules/ding_innskraning/ding_innskraning.module (hunk 26 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module]: Fixed, but it's not used.

File: modules/ding_innskraning/ding_innskraning.module (hunk 28 above)
Comment (xendk) [reply modules/ding_innskraning/ding_innskraning.module:444]: Agreed.

**Review by tripox — APPROVED** [review]
